dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,797 @@
+ * samples to a fallback decoder, if present.
+ *
+ * When a sample is successfully decoded, it is marked as a success in the parent
+ * DecodeResultsPromise. If it fails, it goes to fallback and only if all fallbacks fail, it is
+ * marked in the DecodeResultsPromise as a failure.
+ */
+class ImageDecoder::DecoderWorker {
+ public:
+  /**
+   * @brief Constructs a decoder worker for a given decoder.
+   *
+   * @param owner   - the parent decoder object
+   * @param factory - the factory that constructs the decoder for this worker
+   * @param start   - if true, the decoder is immediately instantiated and the worker thread
+   *                  is launched; otherwise a call to `start` is delayed until the first
+   *                  work that's relevant for this decoder.
+   */
+  DecoderWorker(ImageDecoder *owner, const ImageDecoderFactory *factory, bool start) {
+    owner_ = owner;
+    factory_ = factory;
+    if (start)
+      this->start();
+  }
+  ~DecoderWorker();
+
+  void start();
+  void stop();
+  void add_work(std::unique_ptr<ScheduledWork> work);
+
+  void set_fallback(DecoderWorker *fallback) {
+    fallback_ = fallback;
+  }
+
+  ImageDecoderInstance *decoder(bool create_if_null = true);
+
+ private:
+  std::mutex mtx_;
+  std::condition_variable cv_;
+
+  std::unique_ptr<ScheduledWork> work_;
+  std::thread worker_;
+  bool stop_requested_ = false;
+  std::once_flag started_;
+
+  ImageDecoder *owner_ = nullptr;
+  const ImageDecoderFactory *factory_ = nullptr;
+  std::shared_ptr<ImageDecoderInstance> decoder_;
+  bool produces_gpu_output_ = false;
+
+  /**
+   * @brief Processes a (sub)batch of work.
+   *
+   * The work is scheduled and the results are waited for. Any failed samples will be added
+   * to a fallback work, if a fallback decoder is present.
+   */
+  void process_batch(std::unique_ptr<ScheduledWork> work) noexcept;
+
+  /**
+   * @brief The main loop of the worker thread.
+   */
+  void run();
+
+  // Fallback worker or null, if no fallback is available
+  DecoderWorker *fallback_ = nullptr;
+};
+
+ImageDecoderInstance *ImageDecoder::DecoderWorker::decoder(bool create_if_null) {
+  if (!decoder_) {
+    decoder_ = factory_->Create(owner_->device_id_, owner_->params_);
+    produces_gpu_output_ = factory_->GetProperties().gpu_output;
+  }
+  return decoder_.get();
+}
+
+
+void ImageDecoder::DecoderWorker::start() {
+  std::call_once(started_, [&]() {
+    (void)decoder(true);
+    worker_ = std::thread(&DecoderWorker::run, this);
+  });
+}
+
+ImageDecoder::DecoderWorker::~DecoderWorker() {
+  stop();
+}
+
+void ImageDecoder::DecoderWorker::stop() {
+  if (worker_.joinable()) {
+    {
+      std::lock_guard lock(mtx_);
+      stop_requested_ = true;
+      work_.reset();
+    }
+    cv_.notify_all();
+    worker_.join();
+    worker_ = {};
+  }
+}
+
+void ImageDecoder::DecoderWorker::run() {
+  DeviceGuard dg(owner_->device_id_);
+  while (!stop_requested_) {
```

Comment:
Can we safely read the value without locking the mutex?

Reply:
It's rechecked inside the lock, so nothing wrong should happen - hopefully we won't need to reacquire the lock if stop was requested while processing (but that's rare enough not to bother with).